lib/gc/ExecutionEngine/CPURuntime/Microkernel/BrgemmNaive.cpp

```diff
@@ -0,0 +1,250 @@
+    return storage_ == compare_to.storage_;
+  }
+  bool operator!=(const bf16_t &compare_to) const {
+    return storage_ != compare_to.storage_;
+  }
+  bf16_t(float v) {
+    if (std::isnan(v)) {
+      storage_ = UINT32_C(0x7FC0);
+    } else {
+      caster_t caster;
+      caster.vf = v;
+      uint32_t rounding_bias = ((caster.vl >> 16) & 1) + UINT32_C(0x7FFF);
+      storage_ = static_cast<uint16_t>((caster.vl + rounding_bias) >> 16);
+    }
+  }
+  bf16_t() : storage_(0) {}
+  inline static bf16_t from_storage(uint16_t v) {
+    bf16_t ret;
+    ret.storage_ = v;
+    return ret;
+  }
+};
+
+struct brgemm_params_t {
+  int64_t M, N, K;
+  int64_t LDA, LDB, LDC;
+  int64_t stride_a, stride_b;
+  float beta;
+  int64_t dtypeA, dtypeB;
+  brgemm_params_t() {}
+  brgemm_params_t(int64_t m, int64_t n, int64_t k, int64_t lda, int64_t ldb,
+                  int64_t ldc, int64_t sa, int64_t sb, float b, int64_t da,
+                  int64_t db)
+      : M(m), N(n), K(k), LDA(lda), LDB(ldb), LDC(ldc), stride_a(sa),
+        stride_b(sb), beta(b), dtypeA(da), dtypeB(db) {}
+};
+
+}; // namespace
+
+template <typename C_type>
+static void naive_brgemm_init(const brgemm_params_t &params, void *C,
+                              uint64_t C_offset) {
+  C_type *Cbuf = (C_type *)C;
+  Cbuf += C_offset;
+  for (int i = 0; i < params.M * params.N; i++) {
+    Cbuf[i] = C_type(0);
+  }
+}
+
+static void naive_brgemm_execute_fp32(const brgemm_params_t &params, void *A,
+                                      uint64_t A_offset, void *B,
+                                      uint64_t B_offset, void *C,
+                                      uint64_t C_offset, int num) {
+  float *Abuf = (float *)A;
+  float *Bbuf = (float *)B;
+  float *Cbuf = (float *)C;
+  Abuf += A_offset;
+  Bbuf += B_offset;
+  Cbuf += C_offset;
+  for (int i = 0; i < num; i++) {
+    // a is MxK
+    for (int m = 0; m < params.M; m++) {
+      for (int n = 0; n < params.N; n++) {
+        for (int k = 0; k < params.K; k++) {
+          Cbuf[m * params.LDC + n] +=
+              Abuf[m * params.LDA + k] * Bbuf[k * params.LDB + n];
+        }
+      }
+    }
+    Abuf += params.stride_a;
+    Bbuf += params.stride_b;
+  }
+}
+
+static void naive_brgemm_execute_bf16(const brgemm_params_t &params, void *A,
+                                      uint64_t A_offset, void *B,
+                                      uint64_t B_offset, void *C,
+                                      uint64_t C_offset, int num) {
+  bf16_t *Abuf = (bf16_t *)A;
+  bf16_t *Bbuf = (bf16_t *)B;
+  float *Cbuf = (float *)C;
+  Abuf += A_offset;
+  Bbuf += B_offset;
+  Cbuf += C_offset;
+  for (int i = 0; i < num; i++) {
+    // a is MxK
+    // b is KxNx2k (vnni format)
+    for (int m = 0; m < params.M; m++) {
+      for (int n = 0; n < params.N; n++) {
+        for (int k = 0; k < params.K; k += 2) {
+          Cbuf[m * params.LDC + n] +=
+              Abuf[m * params.LDA + k] * Bbuf[k * params.LDB + 2 * n];
```

Comment:
indexing of `Bbuf` seems incorrect? `k *params.LDB＋n` and `(k+1)*params.LDB +n`?
Otherwise the maximum of this index will be `(params.K-1) * params.LDB + 2(params.N-1)+1`?
Same as int8 subroutine.

Reply:
The calculation is based on VNNI format. The VNNI reordering only change memory layout but not the M, N, K, LDA, LDB values given to BRGEMM (they are the same as plain format), and indexes are set according to it. 
For B buffer, the index is actually `(k / 2) * 2 * params.LDB + 2 * n = k * params.LDB + 2 * n`, and the maximal index should be `(params.K**-2**) * params.LDB + 2(params.N-1)+1` which should be ok.